Sources/NIOSSL/SSLCertificateExtensions.swift

```diff
@@ -0,0 +1,197 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2022 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+@_implementationOnly import CNIOBoringSSL
+@_implementationOnly import CNIOBoringSSLShims
+
+public struct _SSLCertificateExtensions {
+    private enum Storage {
+        final class Deallocator {
+            let reference: OpaquePointer!
+            
+            init(takeOwnershipOf reference: OpaquePointer!) {
+                self.reference = reference
+            }
+            
+            deinit {
+                CNIOBoringSSL_sk_X509_EXTENSION_free(self.reference)
+            }
+        }
+        
+        case owned(Deallocator)
+        case borrowed(reference: OpaquePointer!, owner: AnyObject)
+        
+        init(takeOwnershipOf reference: OpaquePointer!) {
+            self = .owned(.init(takeOwnershipOf: reference))
+        }
+        
+        init(borrowing reference: OpaquePointer!, owner: AnyObject) {
+            self = .borrowed(reference: reference, owner: owner)
+        }
+        
+        /// The owner of the memory to which the reference points
+        var owner: AnyObject {
```

Comment:
Do we need this?

Reply:
Yes because we need to give `_SSLCertificateExtension`, `_SSLCertificateExtension.Data` and `NIOSSLObjectIdentifier` an owner of type `AnyObject`. `_SSLCertificateExtensions` and `_SSLCertificateExtensions.Storage` are not classes and therefore we need this property to get an `AnyObject` we can pass as an owner. We could relax the type constraint to `Any` but I think this will have some negative performance effect (only a theory and not tested) and someone could more easily pass us something wrong.